emacs/config.org
#+TITLE: Emacs Literate Configuration File
#+PROPERTY: header-args:emacs-lisp :tangle ~/code/dotfiles/emacs/init.el

* Initial Setup
** Initialisation Timer

Time exactly how long it takes to load emacs.

#+begin_src emacs-lisp

(defun alex/display-startup-time ()
  (message "Emacs loaded in %s with %d garbage collections."
           (format "%.2f seconds"
                   (float-time
                     (time-subtract after-init-time before-init-time)))
           gcs-done))

(add-hook 'emacs-startup-hook #'alex/display-startup-time)

#+end_src

** Setup Package Repositories

#+begin_src emacs-lisp

(require 'package)
(setq package-archives '(("melpa" . "http://melpa.org/packages/")
			 ("org" . "http://orgmode.org/elpa/")
			 ("elpa" . "http://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
(unless (package-installed-p 'use-package)
 (package-install 'use-package))

(require 'use-package)
(setq use-package-always-ensure t)

#+end_src

** Custom Set Variables

#+begin_src emacs-lisp
;; Auto generated stuff, see my config below
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(company-box-doc-enable nil)
 '(company-box-scrollbar nil)
 '(evil-want-keybinding nil)
 '(lsp-enable-snippet nil)
 '(org-startup-folded t)
 '(org-directory "~/code
/org")
 '(package-selected-packages
   '(toc-org evil-org yasnippet company-tabnine undo-tree nlinum lua-mode typescript-mode web-mode json-mode exec-path-from-shell all-the-icons-dired dired-single evil-magit magit visual-fill-column org-bullets org-mode yasnippet-snippets treemacs-all-the-icons treemacs-projectile lsp-treemacs projectile treemacs-evil python-mode company-lsp ivy-rich eglot lsp-jedi elpy company-box company lsp-mode hydra evil-collection general which-key rainbow-delimiters doom-themes doom-modeline counsel ivy use-package evil))
 '(projectile-mode t nil (projectile)))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(fixed-pitch ((t (:family "Roboto Mono" :height 130))))
 '(org-ellipsis ((t (:foreground "dark gray" :underline nil))))
 '(variable-pitch ((t (:family "Roboto" :height 150)))))
(setq org-agenda-diary-file "~/code/org/agenda.org")

#+end_src

* Improving Emacs

Emacs has certain default behaviours which are just confusing.
These settings are used to make emacs generally more satisfying.

** Override Default Look and Feel

Emacs has a 'default' appearance which needs to end now.

#+begin_src emacs-lisp

(tool-bar-mode -1)
(tooltip-mode -1)
(scroll-bar-mode -1)
(menu-bar-mode -1)
;; Inhibit the splash screen and other default ui elements
(setq inhibit-startup-message t)
(setq initial-scratch-message "; Hi, welcome to emacs :)")
(setq inhibit-startup-echo-area-message "alex")
(set-fringe-mode 0)
;; enable replacing buffer in dired.
(put 'dired-find-alternate-file 'disabled nil)
;; disable question about following symlinks
(setq vc-follow-symlinks nil)
;; stop autosave files being created
(setq auto-save-default nil)
(setq make-backup-files nil)

(use-package term
  :config
  (setq explicit-shell-file-name "bash"))

#+end_src

** Scrolling

These settings help make scrolling more manageable.

#+begin_src emacs-lisp

(setq scroll-conservatively 101)
(setq scroll-margin 1)
(setq redisplay-dont-pause t)
(setq jit-lock-defer-time 0)
;(setq fast-but-imprecise-scrolling t)

(setq mouse-wheel-scroll-amount '(3 ((shift) . 1))) ;; one line at a time
(setq mouse-wheel-progressive-speed nil) ;; don't accelerate scrolling
(setq mouse-wheel-follow-mouse 't) ;; scroll window under mouse

;(require 'smooth-scrolling)
;(use-package smooth-scrolling)
;(smooth-scrolling-mode 1)

#+end_src

* Appearance
** Font Settings

 Create a function that will set all of our faces.

#+begin_src emacs-lisp
  (defun alex/font-faces ()
    (set-face-attribute 'default nil :font "Roboto Mono" :height 130)
    (set-face-attribute 'fixed-pitch nil :font "Roboto Mono" :height 130)
    (set-face-attribute 'variable-pitch nil :font "Roboto" :height 150 :weight 'regular))

#+end_src

Call the function whenever a frame is created. If in daemon mode, select the current frame.

#+begin_src emacs-lisp

  (if (daemonp)
      (add-hook 'after-make-frame-functions
                (lambda (frame)
                 (with-selected-frame frame
                    (alex/font-faces))))
      (alex/font-faces))

#+end_src

** Daemon Configuration

#+begin_src emacs-lisp

(defun alex/daemon-setup (frame)
  (modify-frame-parameters frame
                           '((vertical-scroll-bars . nil)
                             (cursor-color . "dodger blue")
                             (horizontal-scroll-bars . nil))))

(add-hook 'after-make-frame-functions 'alex/daemon-setup)

#+end_src

** Line Numbers

Use the linum package to show a line number column on the left.
This config will leave enough room for a three digit number.

#+begin_src emacs-lisp

(use-package nlinum
  :config
  (setq nlinum-format "%3d "))

(add-hook 'prog-mode-hook (lambda () (nlinum-mode t)))

#+end_src

** Line Wrapping

For =prog-mode=, I prefer lines to be truncated.
I am also changing the default truncation symbol: ~$~ to a space.

#+begin_src emacs-lisp

(setq-default truncate-lines t)
(set-display-table-slot standard-display-table 'truncation 32)

#+end_src

In some other modes I would prefer that the lines are wrapped.

#+begin_src emacs-lisp

(dolist (mode '(org-mode-hook
		term-mode-hook
		eshell-mode-hook))
  (add-hook mode (lambda () (setq visual-line-mode t))))

#+end_src

** Aesthetic Changes

Install all the decorations from doom emacs.
Make sure to use: =M-x RET all-the-icons-install-fonts RET= to install all the icons.
I am also setting the doom modeline height to a sensible value.

#+begin_src emacs-lisp

(use-package all-the-icons)
(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1)
  :custom ((doom-modeline-height 35)))
(setq doom-modeline-enable-word-count nil)
(setq doom-modeline-buffer-encoding nil)
(use-package doom-themes)
(load-theme 'doom-one t)

#+end_src

** Bracket Matching

This package assigns colours to brackets of a certain level.
It makes it much easier to see what is going on, especially in elisp.

#+begin_src emacs-lisp

(use-package rainbow-delimiters
  :hook (prog-mode . rainbow-delimiters-mode))

#+end_src

* Additional Functionality
** Ivy and Counsel

Ivy is a general purpose completion framework for emacs.
Counsel provides an alternative interface for emacs commands making best use of Ivy.
These are the packages which make fuzzy-finding possible while navigating emacs.

#+begin_src emacs-lisp

(use-package counsel
  :bind (("M-x" .  counsel-M-x)
	 ("C-x b" . counsel-ibuffer)
	 ("C-x C-f" . counsel-find-file)
	 :map minibuffer-local-map
	 ("C-r" . 'counsel-minibuffer-history))
  :config
  (setq ivy-initial-inputs-alist nil))

(use-package ivy
  :diminish
  :bind (("C-s" . swiper)
         :map ivy-minibuffer-map
         ("TAB" . ivy-alt-done)
         ("C-l" . ivy-alt-done)
         ("C-j" . ivy-next-line)
         ("C-k" . ivy-previous-line)
         :map ivy-switch-buffer-map
         ("C-k" . ivy-previous-line)
         ("C-l" . ivy-done)
         ("C-d" . ivy-switch-buffer-kill)
         :map ivy-reverse-i-search-map
         ("C-k" . ivy-previous-line)
         ("C-d" . ivy-reverse-i-search-kill))
  :config
  (ivy-mode 1))

(use-package which-key
  :init (which-key-mode)
  :diminish which-key-mode
  :config
  (setq which-key-idle-delay 0.6))

(use-package ivy-rich
  :init
  (ivy-rich-mode 1))

#+end_src

** Shell Path

Make sure that emacs is using the correct path for your preffered shell

#+begin_src emacs-lisp

(use-package exec-path-from-shell
  :ensure t
  :config
  (exec-path-from-shell-initialize))

#+end_src

** Tabbing

Inserting tabs (usually 4 spaces) can be difficult in emacs.
I have bound =C-u= to insert 4 spaces for me.
This is more useful than scrolling upwards as it would do in vim.

#+begin_src emacs-lisp

(setq-default indent-tabs-mode nil)
(setq-default tab-width 4)
(global-set-key (kbd "C-u") 'tab-to-tab-stop)

#+end_src

* Keybindings
** Leader Key

General is a very useful way of setting up a leader key like in vim and other ditrobutions of Emacs.

#+begin_src emacs-lisp

(use-package general
  :after evil
  :config
  (general-create-definer alex/leader-keys
    :keymaps '(normal insert visual emacs)
    :prefix "SPC"
    :global-prefix "C-SPC")

  (alex/leader-keys
   "f" '(:ignore t :which-key "files")
   "ff" '(counsel-find-file :which-key "quick-find")
   "ft" '(treemacs :which-key "toggle treemacs")
   "fs" '(lsp-treemacs-symbols :which-key "toggle lsp symbols")
   "fp" '(projectile-switch-project :which-key "open project")
   "fd" '(dired-jump :which-key "dired-mode"))
  (alex/leader-keys
    "a" '(:ignore t :which-key "actions")
    "at" '(counsel-load-theme :which-key "load-theme")
    "ae" '(eval-buffer :which-key "eval-buffer")
    "al" '(load-file :which-key "load-file"))
  (alex/leader-keys
    "b" '(:ignore t :which-key "buffers")
    "bk" '(kill-this-buffer :which-key "kill this buffer")
    "bK" '(kill-buffer :which-key "kill any buffer")
    "bi" '(ibuffer :which-key "ibuffer")
    "bs" '(counsel-ibuffer :which-key "switch"))
  (alex/leader-keys
    "w" '(:ignore t :which-key "windows")
    "wh" '(evil-window-left :which-key "focus left")
    "wj" '(evil-window-down :which-key "focus down")
    "wk" '(evil-window-up :which-key "focus up")
    "wl" '(evil-window-right :which-key "focus right")
    "wq" '(kill-buffer-and-window :which-key "kill")
    "wf" '(delete-other-windows :which-key "focus this"))
  (alex/leader-keys
    "c" '(:ignore t :which-key "clipboard")
    "cc" '(clipboard-kill-ring-save :which-key "copy")
    "ck" '(clipboard-kill-ring :which-key "cut")
    "cp" '(clipboard-yank :which-key "paste"))
  (alex/leader-keys
    "o" '(:ignore t :which-key "org mode")
    ;"os" '(org-schedule :which-key "schedule")
    ;"od" '(org-deadline :which-key "deadline")
    ;"oa" '(org-agenda :which-key "agenda")
    "ob" '(org-babel-tangle :which-key "export blocks")
    "ol" '(org-store-link :which-key "store link")
    "oi" '(org-insert-last-stored-link :which-key "insert link")
    "or" '(org-mode-restart :which-key "reload"))
  (alex/leader-keys
    "s" '(swiper :which-key "search this file")
    "m" '(magit-status :which-key "magit")
    "j" '(counsel-ibuffer :which-key "switch buffer")
    "k" '(counsel-buffer :which-key "switch buffer"))
  (alex/leader-keys
    "t" '(term :which-key "term"))
  (alex/leader-keys
    "m" '(magit-status :which-key "magit"))
)
;; This package is useful if you want to make a quick menu
(use-package hydra)

#+end_src

** Evil Vim Emulation

The vim keybindings are a must-have feature.
In order to extend the vim behaviour, I have set =ESC= as a global quit key, to make things easy.

#+begin_src emacs-lisp

(global-set-key (kbd "<escape>") 'keyboard-escape-quit)

#+end_src

This is my setup for evil mode.
I have set j and k to move by visual line.
In =prog-mode= this has no effect, but in org mode it makes the keys work like arrow keys.

#+begin_src emacs-lisp

(require 'evil)
;;(setq x-select-enable-clipboard nil)
;;(setq interprogram-cut-function nil)
;;(setq interprogram-paste-function nil)
(setq save-interprogram-paste-before-kill t)
(use-package evil
  :init
  (setq evil-want-integration t)
  (setq evil-want-fine-undo 'fine)
  ;(setq evil-want-C-u-scroll t); Use this option if you want C-u to scroll. I do not.
  :config
  (evil-mode 1)
  (evil-set-undo-system 'undo-tree)
  (define-key evil-insert-state-map (kbd "C-g") 'evil-normal-state)
  (define-key evil-insert-state-map (kbd "C-h") 'evil-delete-backward-char-and-join)
  (evil-global-set-key 'motion "j" 'evil-next-visual-line)
  (evil-global-set-key 'motion "k" 'evil-previous-visual-line)
  (define-key evil-normal-state-map (kbd "C-j") 'counsel-ibuffer)
  (evil-set-initial-state 'messages-buffer-mode 'normal)
  (evil-set-initial-state 'dashboard-mode 'normal)
  (evil-set-initial-state 'term-mode 'normal))

#+end_src

Evil collection is useful for getting vim behaviour in other parts of emacs, besides =prog-mode=.

#+begin_src emacs-lisp

(use-package evil-collection
  :after evil
  :ensure t
  :custom
  (evil-collection-want-unimpaired-p t)
  (evil-collection-company-use-tng t)
  (evil-collection-calendar-want-org-bindings t)
  :config
  (evil-collection-init))

#+end_src

In order to make =C-r= behave like vim, I use undo-tree.
This package is referenced in the evil mode configuration above.

#+begin_src emacs-lisp

(use-package undo-tree
  :after evil
  :config
  (global-undo-tree-mode))

#+end_src

In order to stop the =:q= command closing emacs, rebind using =evil-ex-define-cmd=.

#+begin_src emacs-lisp

(defun alex/evil-write ()
    (interactive)
    (save-buffer)
    (kill-this-buffer))
(evil-ex-define-cmd "wq" 'alex/evil-write)
(evil-ex-define-cmd "q" 'kill-this-buffer)

#+end_src

* General Utilities
** Projectile

Use projectile to display vc projects.
General keybindings are set to show the list of projects.
I clone all of my projects into one folder: =~/code=.
Then i am able to browse all of my projects with one key binding.
When a project is selected, it is opened in a dired buffer, which is useful to keep open while working on the project.

#+begin_src emacs-lisp

(use-package projectile
  :diminish projectile-mode
  :config (projectile-mode)
  :custom ((projectile-completion-system 'ivy))
  :bind-keymap
  ("C-c p" . projectile-command-map)
  :init
  ;; NOTE: Set this to the folder where you keep your Git repos!
  (when (file-directory-p "~/code")
    (setq projectile-project-search-path '("~/code")))
  (setq projectile-switch-project-action #'projectile-dired))

#+end_src

** Magit

Magit is a very useful tool for performing the whole git cycle.
Evil keybindings in this mode come fom evil collection.

#+begin_src emacs-lisp

(use-package magit
  :custom
  (magit-display-buffer-function #'magit-display-buffer-same-window-except-diff-v1))

#+end_src

** Dired

Dired is good by itself, although I do not use it much.
I am also using nice icons in dired.

#+begin_src emacs-lisp

(use-package dired
  :ensure nil
  :commands (dired dired-jump)
  :custom ((dired-listing-switches "-agho --group-directories-first"))
  :config
  (evil-collection-define-key 'normal 'dired-mode-map
    "h" 'dired-single-up-directory
    "l" 'dired-single-buffer))
(use-package dired-single)

(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode))
#+end_src

* Org Mode

Org mode is very useful for notes, programming and both at the same time.

** Org Setup Functions

I copied these functions from the internet.
They set up better looking, variable-pitch fonts in org mode.

#+begin_src emacs-lisp

(defun alex/org-mode-setup ()
  (org-indent-mode)
  (variable-pitch-mode 1) ;; If you want fancy variable width fonts.
  (visual-line-mode 1))

#+end_src

#+begin_src emacs-lisp

(defun alex/org-font-setup ()
  (font-lock-add-keywords 'org-mode
                          '(("^ *\\([-]\\) "
                             (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))

  (dolist (face '((org-level-1 . 1.2)
                  (org-level-2 . 1.1)
                  (org-level-3 . 1.05)
                  (org-level-4 . 1.0)
                  (org-level-5 . 1.1)
                  (org-level-6 . 1.1)
                  (org-level-7 . 1.1)
                  (org-level-8 . 1.1)))
    (set-face-attribute (car face) nil :font "Roboto" :weight 'regular :height 150))

  ;; Ensure that anything that should be fixed-pitch in Org files appears that way
  (set-face-attribute 'org-block nil :foreground nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-code nil   :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-table nil   :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-checkbox nil :inherit 'fixed-pitch))

#+end_src

** Basic Org Setup

#+begin_src emacs-lisp

(require 'org)
(setq org-cycle-separator-lines 2)

(use-package org
  :hook (org-mode . alex/org-mode-setup)
  :config
  ;(setq org-ellipsis " ")
  ;;(setq org-ellipsis " ⤵")
  (setq org-indent-indentation-per-level 2)
  (setq org-hide-emphasis-markers t)
  (setq org-agenda-files '("~/code/org/agenda.org"))
  (alex/org-font-setup))

#+end_src

** Org Mode Extensions

=toc-org= is useful for creating a table of contents for larger files, which also works on github.

#+begin_src emacs-lisp

(use-package toc-org)
(add-hook 'org-mode-hook 'toc-org-mode)

#+end_src

This package allows us to use shortcuts like =<py= to insert code blocks.

#+begin_src emacs-lisp
(with-eval-after-load 'org
  (require 'org-tempo)
  (add-to-list 'org-structure-template-alist '("sh" . "src shell"))
  (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
  (add-to-list 'org-structure-template-alist '("py" . "src python")))

#+end_src

And with =org-babel= we can run those code blocks.

#+begin_src emacs-lisp

(org-babel-do-load-languages
 'org-babel-load-languages
 '((emacs-lisp .t)
   (python . t)))

#+end_src

=org-bullets= replaces the default * symbol with some nice icons.

#+begin_src emacs-lisp

(use-package org-bullets
  :after org
  :hook (org-mode . org-bullets-mode)
  :custom
  (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

#+end_src

To make org more usable, set margins to the left and right.
This prevents text being pushed right up against the side.

#+begin_src emacs-lisp

(defun alex/org-mode-visual-fill ()
  (setq visual-fill-column-width 180
        visual-fill-column-center-text t)
  (visual-fill-column-mode 1))

(use-package visual-fill-column
  :hook (org-mode . alex/org-mode-visual-fill))

#+end_src

* Completion and Highlighting

=lsp-mode= can provide completions and other code actions from a language server running in background.

** LSP Interface Configuration

I don't like lots of elements all at once, so I have disabled a few ui options.

#+begin_src emacs-lisp

(setq lsp-headerline-breadcrumb-enable nil)
(setq lsp-ui-doc-mode 0)
(setq lsp-diagnostics-provider :none)
(setq lsp-signature-auto-activate nil)
(setq lsp-signature-render-documentation nil)
(setq lsp-modeline-code-actions-enable nil)
(setq lsp-modeline-diagnostics-enable nil)
(setq lsp-log-io nil)
(setq lsp-restart 'auto-restart)

#+end_src

** LSP Initialisation

This is the initialisation of =lsp-mode=.

#+begin_src emacs-lisp

(use-package lsp-mode
  :commands (lsp lsp-deferred)
  :init
  (setq lsp-keymap-prefix "C-c l")
  :hook
  (web-mode . lsp-deferred)
  :config
  (lsp-enable-which-key-integration t))
(require 'lsp-mode)
#+end_src

** Company Mode

Company mode provides a nicer graphical interface for completions.
The delay for completions is set to 0s after the first character is typed.
I also use /tab and go/ completion so I do not have to manually select an option.
Completion cycling is also enabled in case I miss the right completion.

#+begin_src emacs-lisp

(use-package company
  :after lsp-mode
  :hook (lsp-mode . company-mode)
  :config
  (setq company-selection-wrap-around t)
  :custom
  (company-minimum-prefix-length 1)
  (company-idle-delay 0.0)
  ; this line is for tab and go completion
  (company-tng-configure-default))
(with-eval-after-load 'company
  ; also use my prefered keys for selection
  (define-key company-active-map (kbd "C-j") #'company-select-next)
  (define-key company-active-map (kbd "C-k") #'company-select-previous))

#+end_src

Company box provides some nice icons next to completions so you can tell rougly what they are.

#+begin_src emacs-lisp
;(use-package company-box ; This package adds some icons in company mode.
;  :after company
;  :diminish
;  :hook (company-mode . company-box-mode))

#+end_src

** Smart Completion

Tabnine is for smart completions.
I am not presently using this completion, as I prefer a default language server.

#+begin_src emacs-lisp

;(use-package company-tabnine :ensure t)
;(require 'company-tabnine)
;(add-to-list 'company-backends #'company-tabnine)
;(add-hook 'prog-mode-hook (lambda () (company-mode t)))
;;(add-hook 'prog-mode-hook (lambda () (company-tabnine t)))

#+end_src

** Start Language Server

Use any language server needed and start =lsp-mode= in certain =prog-mode= buffers.

#+begin_src emacs-lisp

(use-package lsp-jedi
  :ensure t)

(add-hook 'python-mode-hook 'lsp-deferred)
(add-hook 'c++-mode-hook 'lsp-deferred)
(add-hook 'c-mode-hook 'lsp-deferred)

(use-package web-mode
  :mode "\\.js\\'"
  :hook (web-mode . lsp-deferred))

#+end_src

#+begin_src emacs-lisp

      ; lua mode for configuring awesome window manager
 (use-package lua-mode
        :hook (lua-mode-hook . lua-mode))
    ; Haskell mode for configuring xmonad window manager

  (use-package haskell-mode
        :hook (haskell-mode-hook .haskell-mode))

#+end_src

** Treemacs

Use the treemacs file explorer including:

  - projectile support
  - all the icons
  - evil mode keys

#+begin_src emacs-lisp

(use-package lsp-treemacs
  :after lsp)
(use-package treemacs-projectile
  :after (treemacs projectile)
  :ensure t)
(use-package treemacs-all-the-icons
  :after treemacs)
(use-package treemacs-evil
  :after treemacs)

(add-hook 'treemacs-mode-hook (lambda () (treemacs-load-theme "all-the-icons")))
#+end_src


doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
(setq

 doom-font (font-spec :family "Roboto Mono" :weight 'regular :size 16)
 doom-variable-pitch-font (font-spec :family "Roboto" :size 18)
 doom-theme 'doom-one

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.

 display-line-numbers-type nil

;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.

 user-full-name "Alexander Neville"
 user-mail-address "alexander.neville@icloud.com"

)



(setq

 ispell-program-name "hunspell"
 ispell-local-dictionary "en_GB"
 org-directory "~/notes"

 auto-save-default nil
 make-backup-files nil

 mouse-wheel-scroll-amount '(3 ((shift) . 1)) ;; one line at a time
 mouse-wheel-progressive-speed nil ;; don't accelerate scrolling
 mouse-wheel-follow-mouse 't ;; scroll window under mouse
 doom-modeline-height 35
 doom-modeline-enable-word-count nil
 doom-modeline-buffer-encoding nil
 org-cycle-separator-lines 2
 org-startup-folded t
 which-key-idle-delay 0.4
 confirm-kill-emacs nil
 blink-cursor-mode 1

)

;; Disable line highlighting, except in treemacs buffers
(remove-hook 'doom-first-buffer-hook #'global-hl-line-mode)
(add-hook 'treemacs-mode-hook (lambda () (setq hl-line-mode 1)))
;; Remove bracket completion
;;(remove-hook 'doom-first-buffer-hook #'smartparens-global-mode)

(add-hook 'doom-first-buffer-hook #'blink-cursor-mode)
;; Add some margins to text and prog mode buffers


(defun alex/set-up-text-mode ()
  (setq left-margin-width 10)
  (setq right-margin-width 10))

(defun alex/set-up-prog-mode ()
  (setq left-margin-width 2)
  (setq right-margin-width 2))

(add-hook 'text-mode-hook 'alex/set-up-text-mode)
(add-hook 'prog-mode-hook 'alex/set-up-prog-mode)

(custom-set-faces!
  '(font-lock-comment-face :slant italic))

;; Some keybindings i like
(map! :leader
      :desc "switch buffer"
      "b s" #'counsel-ibuffer
      :desc "toggle line highlighting (current buffer)"
      "t h" #'hl-line-mode
      :desc "toggle line highlighting"
      "t H" #'global-hl-line-mode
      :desc "toggle treemacs"
      "f t" #'treemacs
      :desc "Insert line above"
      "i k"   #'+evil/insert-newline-above
      :desc "Insert line below"
      "i j"   #'+evil/insert-newline-below)

(global-set-key (kbd "<escape>") 'keyboard-escape-quit)

(defun alex/org-mode-setup ()
  (font-lock-add-keywords 'org-mode
                          '(("^ *\\([-]\\) "
                             (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))
  (org-indent-mode)
  ;(variable-pitch-mode 1) ;; If you want fancy variable width fonts.
  (visual-line-mode 1))

(use-package org
  :hook (org-mode . alex/org-mode-setup)
  :config
  (setq org-indent-indentation-per-level 2)
  (setq org-hide-emphasis-markers t)
  (setq org-agenda-files '("~/notes/agenda.org")))

;(use-package org-bullets
;  :after org
;  :hook (org-mode . org-bullets-mode)
;  :custom
;  (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

;(use-package! mixed-pitch
;   :hook (org-mode . mixed-pitch-mode)
;   :config
;   (setq mixed-pitch-set-heigth t)
;   (set-face-attribute 'variable-pitch nil :size 18))

(after! org

  (require 'org-tempo)
  (add-to-list 'org-structure-template-alist '("sh" . "src shell"))
  (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
  (add-to-list 'org-structure-template-alist '("py" . "src python"))
  ;(custom-set-faces!
  ;  '(org-document-title :height 1.3)
  ;  '(org-level-1 :inherit outline-1 :weight semi-bold :height 1.3)
  ;  '(org-level-2 :inherit outline-2 :weight semi-bold :height 1.2)
  ;  '(org-level-3 :inherit outline-3 :weight semi-bold :height 1.1)
  ;  '(org-level-4 :inherit outline-4 :weight semi-bold)
  ;  '(org-level-5 :inherit outline-5 :weight semi-bold)
  ;  '(org-level-6 :inherit outline-6 :weight semi-bold)
  ;  '(org-level-7 :inherit outline-7 :weight semi-bold)
  ;  '(org-level-8 :inherit outline-8 :weight semi-bold)
  ;  ;; Ensure that anything that should be fixed-pitch in org buffers appears that way.
  ;  '(org-block nil :foreground nil :inherit 'fixed-pitch)
  ;  '(org-code nil   :inherit '(shadow fixed-pitch))
  ;  '(org-table nil   :inherit '(shadow fixed-pitch))
  ;  '(org-verbatim nil :inherit '(shadow fixed-pitch))
  ;  '(org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
  ;  '(org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
  ;  '(org-checkbox nil :inherit 'fixed-pitch)))
  )



(use-package projectile
  :diminish projectile-mode
  :config (projectile-mode)
  :custom ((projectile-completion-system 'ivy))
  :bind-keymap
  ("C-c p" . projectile-command-map)
  :init
  ;; NOTE: Set this to the folder where you keep your Git repos!
  (when (file-directory-p "~/code")
    (setq projectile-project-search-path '("~/code")))
  (setq projectile-switch-project-action #'projectile-dired))

(require 'lsp-mode)

;; (setq

;;  lsp-headerline-breadcrumb-enable nil
;;  lsp-ui-doc-mode 0
;;  lsp-diagnostics-provider :none
;;  lsp-signature-auto-activate nil
;;  lsp-signature-render-documentation nil
;;  lsp-modeline-code-actions-enable nil
;;  lsp-modeline-diagnostics-enable nil
;;  lsp-log-io nil
;;  lsp-restart 'auto-restart
;;  lsp-enable-snippet nil
;;  company-lsp-enable-snippet nil

;; )

;; (use-package lsp-mode
;;   :commands (lsp lsp-deferred)
;;   :init
;;   (setq lsp-keymap-prefix "C-c l")
;;   :hook
;;   (web-mode . lsp-deferred)
;;   :config
;;   (lsp-enable-which-key-integration t))

(use-package company
  :after lsp-mode
  :hook
  (lsp-mode . company-mode)
  ;(lsp-mode . company-tng-mode)
  :config
  (setq company-selection-wrap-around t)
  :custom
  (company-minimum-prefix-length 1)
  (company-idle-delay 0.0)
  ; this line is for tab and go completion
  ;(company-tng-configure-default))
  )

(with-eval-after-load 'company
   ; also use my prefered keys for selection
   (define-key company-active-map (kbd "C-j") #'company-select-next)
   (define-key company-active-map (kbd "C-k") #'company-select-previous))

;;(add-hook 'python-mode-hook 'lsp-deferred)
